common/map_coordinate.h

```diff
@@ -0,0 +1,231 @@
+struct Transform
+{
+    // Translation
+    const units::length::meter_t tx;
+    const units::length::meter_t ty;
+    const units::length::meter_t tz;
+
+    // Scale
+    const double s;
+
+    // Rotation
+    const units::angle::radian_t rx;
+    const units::angle::radian_t ry;
+    const units::angle::radian_t rz;
+};
+
+//----------------------------------------------------------------------------
+// BoBRobotics::MapCoordinate::Ellipsoid
+//----------------------------------------------------------------------------
+//! Elipsoid associated with a particular datum
+struct Ellipsoid
+{
+    // Major and minor axis
+    const units::length::meter_t a;
+    const units::length::meter_t b;
+
+    // Eccentricity
+    const double f;
+};
+
+//----------------------------------------------------------------------------
+// BoBRobotics::MapCoordinate::WGS84
+//----------------------------------------------------------------------------
+//! WGS84 datum - used for GPS data
+struct WGS84
+{
+    static const Ellipsoid ellipsoid;
+};
+const Ellipsoid WGS84::ellipsoid{units::length::meter_t(6378137), units::length::meter_t(6356752.314245), 1.0 / 298.257223563};
+
+//----------------------------------------------------------------------------
+// BoBRobotics::MapCoordinate::OSGB36
+//----------------------------------------------------------------------------
+// OSGB-36 datum - used for UK OS map coordinates
+struct OSGB36
+{
+    static const Transform transform;
+    static const Ellipsoid ellipsoid;
+};
+const Transform OSGB36::transform{units::length::meter_t(-446.448), units::length::meter_t(125.157), units::length::meter_t(-542.060),
+                                  20.4894, units::angle::arcsecond_t(-0.1502), units::angle::arcsecond_t(-0.2470), units::angle::arcsecond_t(-0.8421)};
+const Ellipsoid OSGB36::ellipsoid{units::length::meter_t(6377563.396), units::length::meter_t(6356256.909), 1.0 / 299.3249646};
+
+// An OS coordinate consisting of an 'easting' and 'northing' distance
+typedef std::tuple<units::length::meter_t, units::length::meter_t> OSCoordinate;
+
+//----------------------------------------------------------------------------
+// BoBRobotics::MapCoordinate::Ellipsoid
+//----------------------------------------------------------------------------
+//! A latitude and longitude relative to a particular datum
+template<typename D>
+struct LatLon
+{
+    typedef D Datum;
+
+    const units::angle::degree_t lat;
+    const units::angle::degree_t lon;
+};
+
+//----------------------------------------------------------------------------
+// BoBRobotics::MapCoordinate::Cartesian
+//----------------------------------------------------------------------------
+//! A cartesian coordinate relative to a particular datum
+template<typename D>
+struct Cartesian
+{
+    typedef D Datum;
+
+    const units::length::meter_t x;
+    const units::length::meter_t y;
+    const units::length::meter_t z;
+};
+
```

Comment:
I like the templating on the datum here -- definitely a nice way to do it

Reply:
Yeah - I was pleased with that bit of C++ inspiration